### ISSUE TITLE
Improve deploy process

### ISSUE THREAD

**Issue by patcon** [edgi-govdata-archiving/100days#29]
Our process got ugly on NameCheap.

There are two options that come to mind:

1. nshapiro suggested namecheap's git deploy method: https://github.com/edgi-govdata-archiving/100days/issues/22
2. Host jekyll site on heroku. there's a way to do this: https://github.com/edgi-govdata-archiving/100days/pull/33
    * SSL instructions: https://devcenter.heroku.com/articles/custom-domains#add-a-custom-domain-with-a-subdomain (down-side is need to redo)
    * heroku has review apps for easy review of changes without building locally. this seems like it would be rad to including other non-tech folks in pull request review process. seems a big plus. https://devcenter.heroku.com/articles/github-integration-review-apps

**Comment by dcwalk** [comment]
Hey, I raised/linked to the doc from @shapironick in #22 ....

Could you unpack the benefits of using heroku? how are review apps different than using github to review changes? (e.g. -- we could request someone else from design look at stuff before deploy if that process needs to be more bridged?)

**Comment by patcon** [comment]
Yes.

And as noted in chat (sorry to fragment), Heroku can be useful for three things:

1. **review of changes.** in-progress changes not yet merged/accepted/approved
2. **view approved pre-release.** viewing the current "approved" final site (ie. master branch), which is ready to go live
3. **deploy release.**

We don't need to use Heroku for that last bit. We could implement a better review now (Heroku), and bookmark sorting a better deploy later. That deploy can either keep using our manual way, or instead investigate namecheap's git deploy #22.

ie. Merging #33 just adds smoother review, and we can revisit deploy later in this issue.

**Comment by patcon** [comment]
Copying @dcwalk's comment from https://github.com/edgi-govdata-archiving/100days/pull/68#issuecomment-331605312

> I just wanna think through out loud who this works for, how we imagine configuring ssh access, etc...
>
> I actually don't think having limited people able to deploy is necessarily a bad thing, and in fact perhaps we should have been a bit better about our review-deploy process this time around (I jumped the gun on one iteration ;).
>
> So questions:
> 1. who would get to? how does this support 'less technical folks'?
> 2. do you see this fitting into the script runner?
> 3. how would people be onboarded to using this?

Would this be a good place to have that convo @dcwalk?

**Comment by patcon** [comment]
Ok, so in regards to deploy, you mentioned that there were some snafu's, and so you're leaning toward limiting deploy rights a bit. I'm wondering whether we could consider a different approach. It's a fairly common devops engineering pattern. The idea is to make deploy of "master" to production fairly simple, but instead controlling highly what gets into "master".

We can do this by turning on [protected branches](https://help.github.com/articles/defining-the-mergeability-of-pull-requests/):
![screenshot of repo protected branch settings](https://imgur.com/Awfjygj.png)

tl;dr these settings force:
1. all commits must go through PR, not master.
2. [Code Owners](https://help.github.com/articles/about-codeowners/) are auto-requested for review, and one approval is required. Owners set by this file:

    ```ini
    # .github/CODEOWNERS
    * @patcon @shaqsingh @dcwalk @suchthis
    ```
3. new commits on a review PR reset its approvals

We can also choose to ensure that status checks must pass in order to merge to "master", and so other other simple assurances can be met with TravisCI or [Dangerfiles](http://danger.systems/ruby/).

So the idea being that if we do some things to be super-rigorous about what gets into master, then it's the conversation about how deploys from master to prod happen is quite a bit less important because "master is always deployable" (which is the mantra that many companies live by).

Curious your thoughts! I realize that much of the above might be familiar to you already, so apologies if I'm rehashing your own know-how @dcwalk :)
